site/publish.sh

```diff
@@ -1,26 +1,132 @@
+# Perform a copy with dereference symlink (object storage do not support symlinks)
+# copy & transform simlinks into referent file/dir
+rsync --archive --checksum --verbose --compress \
+            --copy-links `# derefence symlinks` \
+            --safe-links `# ignore symlinks outside of copied tree` \
+            --stats `# add verbose statistics` \
+            --exclude='updates' \
+            --delete \
+            www2/ www3/
+
+function parallelfunction() {
+    echo "=== parallelfunction: $1"
+
+    case $1 in
+    rsync*)
+        # Push generated index to the production server
+        time rsync --archive --checksum --verbose --compress \
+            --exclude=/updates `# populated by https://github.com/jenkins-infra/crawler` \
+            --delete `# delete old sites` \
+            --stats `# add verbose statistics` \
+            www2/ ${RSYNC_USER}@${UPDATES_SITE}:/var/www/${UPDATES_SITE}
+        ;;
+
+    azsync*)
+        # Sync Azure File Share content using www3 to avoid symlinks
+        time azcopy sync ./www3/ "${UPDATES_FILE_SHARE_URL_AND_PATH}?${UPDATES_FILE_SHARE_QUERY_STRING}" \
+            --recursive=true \
+            --delete-destination=true
+        ;;
+
+    s3sync*)
+        # Sync CloudFlare R2 buckets content excluding 'updates' folder from www3 sync (without symlinks)
+        # as this folder is populated by https://github.com/jenkins-infra/crawler/blob/master/Jenkinsfile
+        time aws s3 sync ./www3/ s3://"${UPDATES_R2_BUCKET}"/ \
+            --no-progress \
+            --no-follow-symlinks \
+            --size-only \
+            --exclude '.htaccess' \
+            --endpoint-url "${UPDATES_R2_ENDPOINT}"
+        ;;
+
+    *)
+        echo -n "Warning: unknown parameter"
+        ;;
+
+    esac
+}
+
+# Export local variables used in parallelfunction
+export UPDATES_SITE
+export RSYNC_USER
+export UPDATES_R2_BUCKET
+export UPDATES_R2_ENDPOINT
+export UPDATES_FILE_SHARE_URL_AND_PATH
+
+# Export function to use it with parallel
+export -f parallelfunction
+
+echo '----------------------- Launch synchronisation(s) -----------------------'
+if [[ $OPT_IN_SYNC_FS_R2 == 'optin' ]]
+then
+    # Sync updates.jenkins.io and azure.updates.jenkins.io
+    parallel --halt-on-error now,fail=1 parallelfunction ::: rsync azsync s3sync
+else
+    # Sync only updates.jenkins.io
+    parallel --halt-on-error now,fail=1 parallelfunction ::: rsync
+
+    # ## If we prefer to avoid parallel when not opt-in, we can replace the previous instruction by the following one:
+    # # push generated index to the production server
+    # rsync --archive --checksum --verbose --compress \
+    #     --exclude=/updates `# populated by https://github.com/jenkins-infra/crawler` \
+    #     --delete `# delete old sites` \
+    #     --stats `# add verbose statistics` \
+    #     ./www2/ ${RSYNC_USER}@${UPDATES_SITE}:/var/www/${UPDATES_SITE}
+fi
+
```

Comment:
I'm annoyed : there are 2 code path with repetition, for a opt-in/opt-out feature it seems like risky.

WDYT about setting the list of elements to run in parallel as part of the if/elese block, and then, outside the block, call the parallel instruction with the list as argument?